api/debug.go

```diff
@@ -95,3 +130,241 @@ func (d *DebugAPI) TraceBlockByHash(
+		emulator.WithChainID(d.config.EVMNetworkID),
+		emulator.WithBlockNumber(head.Number),
+		emulator.WithBlockTime(head.Time),
+	)
+	random := block.PrevRandao
+	stateDB := blockExecutor.StateDB
+	tracer.OnTxStart(
+		&tracing.VMContext{
+			Coinbase:    d.config.Coinbase,
+			BlockNumber: head.Number,
+			Time:        head.Time,
+			Random:      &random,
+			GasPrice:    d.config.GasPrice,
+			ChainConfig: emulatorConfig.ChainConfig,
+			StateDB:     stateDB,
+		},
+		tx,
+		from,
+	)
+
+	err = applyStateOverrides(config, stateDB)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := blockExecutor.Call(from, txEncoded, tracer)
+	if err != nil {
+		return nil, err
+	}
+
+	// call tracer on tx end
+	if tracer.OnTxEnd != nil {
+		tracer.OnTxEnd(result.Receipt(), result.ValidationError)
+	}
+
+	return tracer.GetResult()
+}
+
+func (d *DebugAPI) executorAtBlock(block *models.Block) (*evm.BlockExecutor, error) {
+	blockHeight := block.Height
+	client, err := d.client.GetClientForHeight(blockHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	exeClient, ok := client.(*grpc.Client)
+	if !ok {
+		return nil, fmt.Errorf("could not convert to execution client")
+	}
+
+	ledger, err := evm.NewRemoteLedger(exeClient.ExecutionDataRPCClient(), blockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not create remote ledger for height: %d, with: %w", blockHeight, err)
+	}
+
+	return evm.NewBlockExecutor(
+		block,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
+}
+
+func applyStateOverrides(config *tracers.TraceCallConfig, stateDB types.StateDB) error {
+	if config == nil || config.StateOverrides == nil {
+		return nil
+	}
+
+	diff := *config.StateOverrides
+	for addr, account := range diff {
+		// Override account nonce.
+		if account.Nonce != nil {
+			stateDB.SetNonce(addr, uint64(*account.Nonce))
+		}
+		// Override account(contract) code.
+		if account.Code != nil {
+			stateDB.SetCode(addr, *account.Code)
+		}
+		// Override account balance.
+		if account.Balance != nil {
+			u256Balance, _ := uint256.FromBig((*big.Int)(*account.Balance))
+			currentBalance := stateDB.GetBalance(addr)
+			// If given balance is greater-or-equal than the current balance,
+			// we need to add the diff to the current balance. Otherwise,
+			// we need to sub the diff from the current balance.
+			if u256Balance.Cmp(currentBalance) >= 0 {
+				diff := u256Balance.Sub(u256Balance, currentBalance)
+				stateDB.AddBalance(addr, diff, tracing.BalanceChangeUnspecified)
+			} else {
+				diff := currentBalance.Sub(currentBalance, u256Balance)
+				stateDB.SubBalance(addr, diff, tracing.BalanceChangeUnspecified)
+			}
+		}
+		if account.State != nil && account.StateDiff != nil {
+			return fmt.Errorf("account %s has both 'state' and 'stateDiff'", addr.Hex())
+		}
+		// Replace entire state if caller requires.
+		if account.State != nil {
+			for key, value := range *account.State {
+				stateDB.SetState(addr, key, value)
```

Comment:
What we actually need for this field, is the logic found here: https://github.com/onflow/go-ethereum/blob/master/core/state/statedb.go#L470-L489. But the `stateDB` implementation in `flow-go`, does not have such a method.

Reply:
@ramtinms Would you have any insights on how we can achieve the functionality of the function I listed above?